llms/mlx_lm/models/starcoder2.py

```diff
@@ -0,0 +1,178 @@
+import math
+from dataclasses import dataclass
+from typing import Dict, Optional, Tuple, Union
+
+import mlx.core as mx
+import mlx.nn as nn
+
+from .base import BaseModelArgs
+
+
+@dataclass
+class ModelArgs(BaseModelArgs):
+    model_type: str
+    hidden_size: int
+    num_hidden_layers: int
+    intermediate_size: int
+    num_attention_heads: int
+    num_key_value_heads: int = None
+    max_position_embeddings: int = 16384
+    norm_eps: float = None
+    rms_norm_eps: float = 1e-5
+    vocab_size: int = 49152
+    rope_theta: float = 100000
+
+
+class RMSNorm(nn.Module):
+    def __init__(self, hidden_sizes: int, eps: float = 1e-5):
+        super().__init__()
+        self.weight = mx.ones((hidden_sizes,))
+        self.eps = eps
+
+    def _norm(self, x):
+        return x * mx.rsqrt(x.square().mean(-1, keephidden_sizes=True) + self.eps)
+
+    def __call__(self, x):
+        output = self._norm(x.astype(mx.float32)).astype(x.dtype)
+        return self.weight * output
+
+
+class Attention(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+
+        dim = args.hidden_size
+        self.n_heads = n_heads = args.num_attention_heads
+        self.n_kv_heads = n_kv_heads = args.num_key_value_heads
+
+        self.repeats = self.n_heads // self.n_kv_heads
+
+        head_dim = args.hidden_size // args.num_attention_heads
+        self.scale = head_dim**-0.5
+
+        self.wq = nn.Linear(dim, n_heads * head_dim, bias=False)
```

Comment:
Starcoder2 uses the standard HF attention layer name, so you may need to update the q,k,v,o layer name here as well.
https://github.com/huggingface/transformers/blob/main/src/transformers/models/starcoder2/modeling_starcoder2.py#L211

Reply:
thanks but Mistral also uses them https://github.com/huggingface/transformers/blob/f1b1379f37c6b9626bb1c795d89be4c0a606f957/src/transformers/models/mistral/modeling_mistral.py#L227

and in MLX implentation of Mistral have same name
https://github.com/ml-explore/mlx-examples/blob/3acc1ec84e838d1070b04ab77b76e31176955a8d/llms/mistral/mistral.py#L55C13-L55C14

or are you referring to sth else

Reply:
Thank you very much for all the help